### ISSUE TITLE
Stack overflow in VRTK_PlayerClimb -> OnTeleport

### ISSUE THREAD

**Issue by Prime6** [thestonefox/VRTK#735]
When you have a lot of prefabs (e.q. > 65 buildings) with multiple 3d objects inside (cubes, cylinders...) and you (e.g.) walk up the stairs of a building, the VRTK_PlayerClimb teleport function (line 200++) crashes with a stack overflow exception. This makes it difficult to procedurally generate climbing environments with hundreds of objects. (edit: even 65 give problems...)

**Comment by thestonefox** [comment]
I can't see why you're getting a stack overflow just because you have lots of objects. The player climb script should only be dealing with one object at a time.

Can you provide a way to recreate the issue?

**Comment by Prime6** [comment]
I also don't understand why this only happens with lots of objects. Sure, I will provide the scene asset + prefabs and some more Informationen later, because I'm at work right now. I'm using the latest Unity version 5.4.3.

**Comment by Prime6** [comment]
Hi, so this is getting weirder... if I disable VRTK_PlayerClimb (& VRTK_Height Adjust Teleport) when playing the Game, I can only move with ArmSwinger (locomotion): this is ok. 

But when I enable these scripts while playing through Unity, it is working with 128 buildings. If I enable those scripts before starting the game, it crashes right at the beginning of moving with bezier curves (VRTK Teleport feature with controllers). 

[Climb_Assets.zip](https://github.com/thestonefox/VRTK/files/613729/Climb_Assets.zip)
I have attached a ZIP-File with a basic setup. Load Level1 from the Scene folder. Click on the "Ground" entity and set "max_buildings" to 128. Play around with the scripts attached to [Camera Rig].

Unfortunately I cannot reproduce the stack overflow log entry ... dunno why. But maybe you have a clue what is going on with moving / teleporting / position correction.

**Comment by Prime6** [comment]
I finally managed to reproduce the stack overflow bug with my setup. See attached file for proof.  
![stack_overflow_player_climb](https://cloud.githubusercontent.com/assets/19248333/20633155/de586476-b343-11e6-8dde-d0c0b9f1299a.PNG)

How to reproduce:
* set max_buildings to 65
* enable all scripts attached to the [Camera Rig] entity (scripts on controllers as well!)
* start game and move with controller teleport (bezier pointer; touchpad button) or use ArmSwinger and touch a building with your "head" / move upstairs the wooden basement of the building.

Interesting fact:
If I set max_buildings to 64 there is now bug for me...

**Comment by thestonefox** [comment]
I can get your scene to crash. But I think the issue may be elsewhere. I'm going to see if I can re-create the issue in a new scene.

**Comment by thestonefox** [comment]
I can get an example scene to crash unity without using your example scene. I'll investigate it further until I find the issue,

**Comment by thestonefox** [comment]
@Prime6  try this:

Disable the `height adjust teleporter` script in your editor (before clicking play scene).

Then play the scene with it disabled. wait a couple of seconds and then enable the `height adjust teleporter` script then try teleporting.

Do you still get an error or a crash then?

I've found that doesn't crash things when i do that. Hopefully it will help narrow the issue down!

**Comment by thestonefox** [comment]
Yeah I can confirm that in my example scene if I disable `Height Adjust Teleporter` then enable it at runtime things work fine. Also, if I have 64 distinct items it works, go to 65 items and I need to disable it before playing the scene...

This is my example scene for reference:

![image](https://cloud.githubusercontent.com/assets/1029673/20635227/4e8926dc-b352-11e6-85ca-1182519bf67e.png)


It's definitely something to do with the Height Adjust Teleporter!

**Comment by Prime6** [comment]
Hey, good to see that you can re-create the bug with my scene and another scene. Really strange that it is working with 64 but not with 65. 

I will try your test case out later this day when I'm back at my PC. It's Interesting to see that the Height Adjust Teleporter is causing the crash and not the PlayerClimb as I suspected and showed with a screenshot. Could this be a dependency bug? 

Overall it's looking for me like some kind of race condition bug, so something is initialized a bit later - that is why the activation later on is working. What do you think?

**Comment by thestonefox** [comment]
@Prime6 yeah it could be a race condition.

It's like the listener isn't getting set up in time, perhaps it's a memory leak within unity itself?!

I may try and recreate it with much less code and if I can look to submit to unity as a potential bug.

**Comment by Prime6** [comment]
@thestonefox I have just tested your testcase. It is also working for me if I disable the Height Adjust Teleporter right before playing. When I enable it while playing I can move around and do all the things I want. So to answer your question: no I don't get any crashs if I do it the way you said. 

Interesting fact: I can only teleport around with the bezier pointers (script on controllers) when the Height Adjust Teleporter script is activated - is that the way it is meant to be? Moving with ArmSwinger is no problem with or without the script.

Regarding the race condition: is it possible to initialize some instances in "Awake" or "Late update" ? Maybe that could solve the problem. Or some kind of coroutine with a delay of a few seconds? Just some ideas...

Let me know what you think and what you have figured out while re-creating the bug with less code. Best regards.

**Comment by thestonefox** [comment]
@Prime6 yes the height adjust teleporter has to be active as that is the thing that is actually doing the teleport.

If you could still teleport with the script deactivated then it would be the incorrect behaviour as deactivated scripts shouldn't really be providing a function, otherwise you can't toggle behaviour during runtime by toggling scripts.

I also tried running the initialisation in co-routines at the end of frame or using Invoke and waiting a fraction of a second, it didn't help :(

**Comment by Prime6** [comment]
@thestonefox @height adjust teleporter: ok, I undestand that. 

Yes, I agree, this would be really strange if disabled scripts would provide functionality.

Hmm... that is too bad :( - how about activating the script via code after some time? Is that possible - haven't checked the internet for this.

There has to be a workaround :) !

Just let me know if you need help. I really appreciate your work and I am willing to help.

( Btw: I tried my demo level with round about 512 buildings and the framerate dropped a lot! After some investigation in the Profiler I saw that Interactable Object.FixedUpdate() is causing the framedrop. It's called for every climable object in the scene and thats quite a huge number. Sorry, If this is offtopic. )

**Comment by Prime6** [comment]
BTW: I'm having huge performance framedrops in case of more than 200 buildings in a scene. All buildings contain multiple objects, which are all climable. The low fps is caused by VRTK_Climable.FixedUpate(), which is executed a lot per second. You can try it out on your own with my demo Scene and monitor this with the performance monitor.

**Comment by thestonefox** [comment]
@Prime6 there is an open issue related to the Interactable Object updates:

https://github.com/thestonefox/VRTK/issues/733

The plan is to auto disable the interactable object scripts when you're not interacting with them so their update routines won't be called.

**Comment by Prime6** [comment]
@thestonefox oh! Thank you, I will have a look into that :-)

Do you have any updates on the teleport / moving issue?

**Comment by thestonefox** [comment]
@Prime6 I've not had chance to look into a solution for the teleport issue yet.

However, for a workaround you can disable the teleport script, then just enable it via another script that just enables it after a second or something, like so:

```
void OnEnable() {
  Invoke("EnableTeleport", 1f);
}

void EnableTeleport() {
  GetComponent<VRTK_HeightAdjustTeleport>().enable;
}
```

**Comment by Prime6** [comment]
@thestonefox okay, no problem at all. 

I will try out your example code asap and give feedback.

**Comment by Prime6** [comment]
@thestonefox Hi. Sorry for the late reply. I'm back with good news: If I enable the teleporter script with the invoke method (example scene has 15 seconds delay) it's working :)!

I'll also try out your interactable object update and let you know my results.

Thanks!

**Comment by Prime6** [comment]
@thestonefox I'm back for a (late) reply: I can not get your branch to work in my environment. So i can not check the performance increase with disabled interactable objects. That is why I am waiting for your PR release ;-).

**Comment by thestonefox** [comment]
This has all been merged into master branch now:
https://github.com/thestonefox/VRTK/commit/22c1573e69b0e30a9ccbe17d2391d93931978b65

**Comment by Prime6** [comment]
Yay :D ! Have to try it out. Thanks :)
